xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/LocalMUCRoom.java

```diff
@@ -792,41 +792,47 @@ private void sendInitialPresences(LocalMUCRole joinRole) {
-        // Create a proxy for the occupant that joined the room from another cluster node
-        RemoteMUCRole joinRole = new RemoteMUCRole(mucService, event);
-        JID bareJID = event.getUserAddress().asBareJID();
-        String nickname = event.getNickname();
-        List<MUCRole> occupants = occupantsByNickname.get(nickname.toLowerCase());
-        // Do not add new occupant with one with same nickname already exists
-        if (occupants == null) {
-            occupants = new ArrayList<>();
-            occupantsByNickname.put(nickname.toLowerCase(), occupants);
-        } else {
-            // sanity check; make sure the nickname is owned by the same JID
-            if (occupants.size() > 0) {
-                JID existingJID = occupants.get(0).getUserAddress().asBareJID();
-                if (!bareJID.equals(existingJID)) {
-                    Log.warn(MessageFormat.format("Conflict detected; {0} requested nickname '{1}'; already being used by {2}", bareJID, nickname, existingJID));
-                    return;
+        lock.writeLock().lock();
+        try {
+            // Create a proxy for the occupant that joined the room from another cluster node
+            RemoteMUCRole joinRole = new RemoteMUCRole(mucService, event);
+            JID bareJID = event.getUserAddress().asBareJID();
+            String nickname = event.getNickname();
+            List<MUCRole> occupants = occupantsByNickname.get(nickname.toLowerCase());
+            // Do not add new occupant with one with same nickname already exists
+            if (occupants == null) {
+                occupants = new ArrayList<>();
+                occupantsByNickname.put(nickname.toLowerCase(), occupants);
+            } else {
+                // sanity check; make sure the nickname is owned by the same JID
+                if (occupants.size() > 0) {
+                    JID existingJID = occupants.get(0).getUserAddress().asBareJID();
+                    if (!bareJID.equals(existingJID)) {
+                        Log.warn(MessageFormat.format("Conflict detected; {0} requested nickname '{1}'; already being used by {2}", bareJID, nickname, existingJID));
+                        return;
+                    }
                 }
             }
-        }
-        // Add the new user as an occupant of this room
-        occupants.add(joinRole);
-        // Update the tables of occupants based on the bare and full JID
-        List<MUCRole> list = occupantsByBareJID.get(bareJID);
-        if (list == null) {
-            list = new ArrayList<>();
-            occupantsByBareJID.put(bareJID, list);
-        }
-        list.add(joinRole);
-        occupantsByFullJID.put(event.getUserAddress(), joinRole);
+            // Add the new user as an occupant of this room
+            occupants.add(joinRole);
+            // Update the tables of occupants based on the bare and full JID
+            List<MUCRole> list = occupantsByBareJID.get(bareJID);
```

Comment:
Not as a result of your changes, but as you're here, lines 819->825 can be replaced with simply

`occupantsByBareJID.computeIfAbsent(bareJID, jid->new ArrayList<>()).add(joinRole);`

Reply:
done